src/Datadog.Trace/HttpOverStreams/DatadogHttpClient.cs

```diff
@@ -57,27 +58,83 @@ private HttpResponse ReadResponse(Stream responseStream)
             const int statusCodeStart = 9;
             const int statusCodeEnd = 12;
             const int startOfReasonPhrase = 13;
+            const int bufferSize = 10;
 
             // TODO: Get this from StringBuilderCache after we determine safe maximum capacity
             var stringBuilder = new StringBuilder();
 
-            var chArray = new byte[1];
-            void GoNextChar()
+            var chArray = new byte[bufferSize];
+
+            async Task GoNextChar()
             {
-                streamPosition++;
-                chArray[0] = (byte)responseStream.ReadByte();
+                var bytesRead = await responseStream.ReadAsync(chArray, offset: 0, count: 1).ConfigureAwait(false);
+                if (bytesRead == 0)
+                {
+                    throw new InvalidOperationException($"Unexpected end of stream at position {streamPosition}");
+                }
+
                 currentChar = Encoding.ASCII.GetChars(chArray)[0];
+                streamPosition++;
+            }
+
+            async Task SkipUntil(int requiredStreamPosition)
+            {
+                var advanceCount = requiredStreamPosition - streamPosition;
+                // Not required in release mode, as should only arise from programming error
+                System.Diagnostics.Debug.Assert(advanceCount > 0, "RequiredStreamPosition should be greater than 0");
+                System.Diagnostics.Debug.Assert(advanceCount <= bufferSize, "RequiredStreamPosition should be less than buffer size");
+                var totalBytesRead = 0;
+
+                while (totalBytesRead < advanceCount)
+                {
+                    var bytesRead = await responseStream.ReadAsync(chArray, offset: 0, count: advanceCount).ConfigureAwait(false);
+                    if (bytesRead == 0)
+                    {
+                        throw new InvalidOperationException($"Unexpected end of stream at position {streamPosition}");
+                    }
+
+                    totalBytesRead += bytesRead;
+                }
```

Comment:
I think the `advanceCount <= bufferSize` check should also be there in release (especially since we run the tests in release mode now )
Though you shouldn't even need it. If you make sure that you read `Math.Min(advanceCount, bufferSize)`, your code should work as-is with any value of `advanceCount`
(it would just require an extra bit of logic to get the position of the `currentChar` after the reading loop)

Reply:
fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1218/commits/c8b9639fdd7acf8ab8c0c774736c8cc661a1bbee

> (it would just require an extra bit of logic to get the position of the currentChar after the reading loop)

This should have been in there anyway, to handle the case (where fewer than `advanceCount` were returned in the stream)
so you inadvertently fixed a bug too :smile: